pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -65,6 +67,10 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, prox
 		if configgen.CanUsePrecomputedCDS(proxy) {
 			if sidecarScope != nil && sidecarScope.XDSOutboundClusters != nil {
 				clusters = append(clusters, sidecarScope.XDSOutboundClusters...)
+				// For locality loadbalancing
+				for _, cluster := range clusters {
+					ApplyLocalitySetting(proxy, cluster, push)
```

Comment:
wont this modify the cache's copy directly? and this will cause data races when there are concurrent CDS computations for multiple proxies as they all will modify same xdsoutboundClusters.

I was thinking that it would be easier to have a XDSOutboundClusters per source locality, in the sidecar scope object. based on the proxy Node's locality, we can select the appropriate cached version. This will become useful to EDS as well

Reply:
yes, in EDS I have done such thing. 
Why i did not do it here is that:  We do not have many clusters with LoadAssignment.  This percentage should be very small.  If we do cache per locality, then the cache I think will take up several times more memory.

> wont this modify the cache's copy directly?
This is a good point. I found too, can prevent race by shallow copy `LocalityLbEndpoints`, because we only need update `LocalityLbEndpoints.LoadBalancingWeight` or `LocalityLbEndpoints.Priority`

Reply:
FYI: https://github.com/istio/istio/pull/10720/commits/f0c61ddaaa67bf9cf311a0038ca7659294b31807#diff-f507fb55b7edf4a03d3c7e9aebf64a1eR73